specs/sharding/beacon-chain.md

```diff
@@ -613,13 +730,39 @@ def process_shard_header(state: BeaconState, signed_header: SignedShardBlobHeade
         == bls.Pairing(body_summary.commitment.point, G2_SETUP[-body_summary.commitment.length])
     )
 
+    # Charge EIP 1559 fee, builder pays for opportunity, and is responsible for later availability,
+    # or fail to publish at their own expense.
+    samples = blob_summary.commitment.length
+    # TODO: overflows, need bigger int type
+    max_fee = blob_summary.max_fee_per_sample * samples
+
+    # Builder must have sufficient balance, even if max_fee is not completely utilized
+    assert state.blob_builder_balances[header.builder_index] >= max_fee
+
+    base_fee = state.shard_sample_price * samples
+    # Base fee must be paid
+    assert max_fee >= base_fee
```

Comment:
I wonder if a more neutral assertion is:
```python
 assert blob_summary.max_fee_per_sample >= state.shard_sample_price
```
Basically nothing will work unless builder pays at least the sample price in state. Can also move such assertion a few lines up too

Reply:
Good point, we can factor out `samples` and do another earlier comparison. But if we get the precision (uint size) right to avoid overflows, it doesn't make a difference. I'm keeping this as-is, and will address fee precision and corresponding fee asserts in a future PR.